app/login.go

```diff
@@ -158,11 +158,13 @@ func (a *App) DoLogin(w http.ResponseWriter, r *http.Request, user *model.User,
 	var err *model.AppError
 	if session, err = a.CreateSession(session); err != nil {
 		err.StatusCode = http.StatusInternalServerError
-		return nil, err
+		return err
 	}
 
 	w.Header().Set(model.HEADER_TOKEN, session.Token)
 
+	a.Session = *session
+
 	if pluginsEnvironment := a.GetPluginsEnvironment(); pluginsEnvironment != nil {
 		a.Srv.Go(func() {
 			pluginContext := a.PluginContext()
```

Comment:
Related to this change, we now guarantee that the plugin context actually has the session. Could we add a unit test for this case to keep it that way? (Ideally something that would have failed before, but I realize that's hard given the goroutine.)

Reply:
I will look into this.

Ideally we want to build the app with the `-race` flag and make sure that it doesn't report anything. Obviously we are not there yet since there are several races still in the code.  
Also, from my prior investigation I've seen tests introducing new kinds of race conditions that are not present when running the application.